### ISSUE TITLE
Server Error with 2.13.1 (dirty .war, 2.13.1 plugin)

### ISSUE THREAD

**Issue by nicholaschum** [davido/gerrit-oauth-provider#71]
Whenever trying to connect from a new GitHub account, I happen to get this error while trying to log in. Unfortunately I was not able to find a fix in the other open issues/closed issues.

`[2016-10-08 16:00:29,890] [HTTP-64] ERROR com.google.gerrit.pgm.http.jetty.HiddenErrorHandler : Error in GET /oauth?code=9de63243bfe864d7490c&state=h08cQlFkv6uufZ0hY7Zx6-RsSmd6nTHf__JJSJIfKQI
org.scribe.exceptions.OAuthException: Response body is incorrect. Can't extract a token from this: 'error=bad_verification_code&error_description=The+code+passed+is+incorrect+or+expired.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fv3%2Foauth%2F%23bad-verification-code'
	at org.scribe.extractors.TokenExtractor20Impl.extract(TokenExtractor20Impl.java:32)
	at org.scribe.oauth.OAuth20ServiceImpl.getAccessToken(OAuth20ServiceImpl.java:37)
	at com.googlesource.gerrit.plugins.oauth.GitHubOAuthService.getAccessToken(GitHubOAuthService.java:116)
	at com.google.gerrit.httpd.auth.oauth.OAuthSession.login(OAuthSession.java:98)
	at com.google.gerrit.httpd.auth.oauth.OAuthWebFilter.doFilter(OAuthWebFilter.java:109)
	at com.google.gwtexpui.server.CacheControlFilter.doFilter(CacheControlFilter.java:73)
	at com.google.gerrit.httpd.RunAsFilter.doFilter(RunAsFilter.java:122)
	at com.google.gerrit.httpd.RequestMetricsFilter.doFilter(RequestMetricsFilter.java:60)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy$1.doFilter(AllRequestFilter.java:136)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy.doFilter(AllRequestFilter.java:138)
	at com.google.gerrit.httpd.RequestContextFilter.doFilter(RequestContextFilter.java:75)
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:135)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
`

**Comment by nicholaschum** [comment]
Changing [auth] to OPENID worked as suggested in https://github.com/davido/gerrit-oauth-provider/issues/70#issuecomment-252023116

**Comment by zzzeek** [comment]
I'm also having this problem and I fail to see how using two separate auth providers is a solution.

**Comment by zzzeek** [comment]
downgraded back to gerrit 2.12 and the plugin, everything works.  definitely a regression of some kind.  Here's the log output in case someone looks at it:

```
[2016-10-19 16:55:12,849] [main] INFO  com.google.gerrit.pgm.Daemon : Gerrit Code Review 2.13.1 ready
[2016-10-19 16:56:23,631] [HTTP-59] ERROR com.google.gerrit.httpd.auth.oauth.OAuthSession : Illegal request state 'sI4sHEpX8UI6RO7ozmfZ6WaEy_6AOLGlV9jYUXdBAKI' on OAuthProtocol OAuthSession [token=null, user=null]
[2016-10-19 16:56:24,159] [HTTP-60] WARN  org.eclipse.jetty.servlet.ServletHandler : /oauth
org.scribe.exceptions.OAuthException: Response body is incorrect. Can't extract a token from this: 'error=bad_verification_code&error_description=The+code+passed+is+incorrect+or+expired.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fv3%2Foauth%2F%23bad-verification-code'
	at org.scribe.extractors.TokenExtractor20Impl.extract(TokenExtractor20Impl.java:32)
	at org.scribe.oauth.OAuth20ServiceImpl.getAccessToken(OAuth20ServiceImpl.java:37)
	at com.googlesource.gerrit.plugins.oauth.GitHubOAuthService.getAccessToken(GitHubOAuthService.java:116)
	at com.google.gerrit.httpd.auth.oauth.OAuthSession.login(OAuthSession.java:101)
	at com.google.gerrit.httpd.auth.oauth.OAuthWebFilter.doFilter(OAuthWebFilter.java:109)
	at com.google.gerrit.httpd.RequireSslFilter.doFilter(RequireSslFilter.java:77)
	at com.google.gwtexpui.server.CacheControlFilter.doFilter(CacheControlFilter.java:73)
	at com.google.gerrit.httpd.RunAsFilter.doFilter(RunAsFilter.java:122)
	at com.google.gerrit.httpd.RequestMetricsFilter.doFilter(RequestMetricsFilter.java:60)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy$1.doFilter(AllRequestFilter.java:136)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy.doFilter(AllRequestFilter.java:105)
	at com.google.gerrit.httpd.RequestContextFilter.doFilter(RequestContextFilter.java:75)
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:135)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)
[2016-10-19 16:56:24,161] [HTTP-60] ERROR com.google.gerrit.pgm.http.jetty.HiddenErrorHandler : Error in GET /oauth?code=0e9ae97810ab06c1eb67&state=SUMnuA83Uj5H59Xk8mrKxqLP7HRBYWr9q85nhXyWJs8
org.scribe.exceptions.OAuthException: Response body is incorrect. Can't extract a token from this: 'error=bad_verification_code&error_description=The+code+passed+is+incorrect+or+expired.&error_uri=https%3A%2F%2Fdeveloper.github.com%2Fv3%2Foauth%2F%23bad-verification-code'
	at org.scribe.extractors.TokenExtractor20Impl.extract(TokenExtractor20Impl.java:32)
	at org.scribe.oauth.OAuth20ServiceImpl.getAccessToken(OAuth20ServiceImpl.java:37)
	at com.googlesource.gerrit.plugins.oauth.GitHubOAuthService.getAccessToken(GitHubOAuthService.java:116)
	at com.google.gerrit.httpd.auth.oauth.OAuthSession.login(OAuthSession.java:101)
	at com.google.gerrit.httpd.auth.oauth.OAuthWebFilter.doFilter(OAuthWebFilter.java:109)
	at com.google.gerrit.httpd.RequireSslFilter.doFilter(RequireSslFilter.java:77)
	at com.google.gwtexpui.server.CacheControlFilter.doFilter(CacheControlFilter.java:73)
	at com.google.gerrit.httpd.RunAsFilter.doFilter(RunAsFilter.java:122)
	at com.google.gerrit.httpd.RequestMetricsFilter.doFilter(RequestMetricsFilter.java:60)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy$1.doFilter(AllRequestFilter.java:136)
	at com.google.gerrit.httpd.AllRequestFilter$FilterProxy.doFilter(AllRequestFilter.java:105)
	at com.google.gerrit.httpd.RequestContextFilter.doFilter(RequestContextFilter.java:75)
	at com.google.inject.servlet.ManagedFilterPipeline.dispatch(ManagedFilterPipeline.java:120)
	at com.google.inject.servlet.GuiceFilter.doFilter(GuiceFilter.java:135)
	at org.eclipse.jetty.servlet.ServletHandler$CachedChain.doFilter(ServletHandler.java:1652)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:585)
	at org.eclipse.jetty.server.session.SessionHandler.doHandle(SessionHandler.java:221)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1127)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:515)
	at org.eclipse.jetty.server.session.SessionHandler.doScope(SessionHandler.java:185)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1061)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:97)
	at org.eclipse.jetty.server.Server.handle(Server.java:499)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:311)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:257)
	at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
	at java.lang.Thread.run(Thread.java:745)

```

**Comment by nicholaschum** [comment]
@zzzeek it isn't a solution. I ultimately downgraded my gerrit server instead because it wasn't working properly.

**Comment by davido** [comment]
OK, any specific insruction how to reproduce it?

**Comment by davido** [comment]
OK, I see now what is going on. This is another regression in OAuth extension point in 2.13, *sigh*. I will upload the fix upstream in a moment: [1].

* [1] https://bugs.chromium.org/p/gerrit/issues/detail?id=4797

**Comment by davido** [comment]
Fixed in: [1].

* [1] https://gerrit-review.googlesource.com/#/c/89329

**Comment by davido** [comment]
I've updated the emergency fix and uploaded it in this release: [1].

* [1] https://github.com/davido/gerrit-oauth-provider/releases/tag/v2.13.1

**Comment by nicholaschum** [comment]
Great. Thanks @davido!

**Comment by zzzeek** [comment]
thanks very much!  i'm back on 2.12 at the moment but I'll try out 2.13 again as part of my long slog towards upgrading my environment...

**Comment by davido** [comment]
So, the 2.13.2 was released today morning: [1] and should include all fixes for regression in the OAuth extension point. I'm going to remove emergency fix from the plugin 2.13.1 release.

* [1] https://groups.google.com/d/topic/repo-discuss/W5hoCVA7oVE/discussion

**Comment by davido** [comment]
2.13.1 release page was updated: [1].

* [1] https://github.com/davido/gerrit-oauth-provider/releases/tag/v2.13.1
